### PR TITLE
add deregistration for paths

### PR DESCRIPTION
**Pull request by deads2k**
Aggregation and TPRs require the ability to de-register endpoints.  This adds the capability and makes it threadsafe.

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -241,20 +241,28 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService, de
...
-	s.GenericAPIServer.HandlerContainer.ServeMux.Handle(groupPath, groupDiscoveryHandler)
-	s.GenericAPIServer.HandlerContainer.ServeMux.Handle(groupPath+"/", groupDiscoveryHandler)
+	s.GenericAPIServer.FallThroughHandler.Handle(groupPath, groupDiscoveryHandler)
+	s.GenericAPIServer.FallThroughHandler.UnlistedHandle(groupPath+"/", groupDiscoveryHandler)
 	s.handledGroups.Insert(apiService.Spec.Group)
 }
 
-// RemoveAPIService removes the APIService from being handled.  Later on it will disable the proxy endpoint.
+// RemoveAPIService removes the APIService from being handled.
 // Right now it does nothing because our handler has to properly 404 itself since muxes don't unregister
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: remove this comment

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -241,20 +241,28 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService, de
...
 	// aggregation is protected
-	s.GenericAPIServer.HandlerContainer.ServeMux.Handle(groupPath, groupDiscoveryHandler)
-	s.GenericAPIServer.HandlerContainer.ServeMux.Handle(groupPath+"/", groupDiscoveryHandler)
+	s.GenericAPIServer.FallThroughHandler.Handle(groupPath, groupDiscoveryHandler)
+	s.GenericAPIServer.FallThroughHandler.UnlistedHandle(groupPath+"/", groupDiscoveryHandler)
 	s.handledGroups.Insert(apiService.Spec.Group)
 }
 
-// RemoveAPIService removes the APIService from being handled.  Later on it will disable the proxy endpoint.
+// RemoveAPIService removes the APIService from being handled.
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: doc that this method is not thread-safe, call in the same thread as AddAPIService

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/helpers.go [hunk 3]
```diff
@@ -57,3 +59,8 @@ func (s ByPriority) Less(i, j int) bool {
 	}
 	return s[i].Spec.Priority < s[j].Spec.Priority
 }
+
+func APIServiceNameToGroupVersion(apiServiceName string) schema.GroupVersion {
+	tokens := strings.SplitN(apiServiceName, ".", 2)
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/helpers.go]: err if len(tokens) != 2

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis.go [hunk 4]
```diff
@@ -75,11 +59,6 @@ var discoveryGroup = metav1.APIGroup{
...
 
 func (r *apisHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	// if the URL is for OUR api group, serve it normally
-	if strings.HasPrefix(req.URL.Path+"/", "/apis/"+apiregistrationapi.GroupName+"/") {
-		r.delegate.ServeHTTP(w, req)
-		return
-	}
 	// don't handle URLs that aren't /apis
 	if req.URL.Path != "/apis" && req.URL.Path != "/apis/" {
 		r.delegate.ServeHTTP(w, req)
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis.go]: unrelated, but I just noticed we lost `content-type: application/json` on the following URLs:
* `https://<server>/apis`
* `https://<server>/apis/`
* `https://<server>/apis/extensions`
* `https://<server>/apis/extensions/`

Oddly, the same urls against the unsecured port still return `content-type: application/json`. I didn't expect the handler (once you got through the outer filters) to be different.

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/helpers.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/helpers.go]: > err if len(tokens) != 2

I'll add doc for it to be a valid name.  It's already enforced in API validation.

**Pull request merged**